Sources/Datadog/Core/System/CarrierInfoProvider.swift

```diff
@@ -66,88 +66,147 @@ extension CarrierInfo.RadioAccessTechnology {
 
-    var current: CarrierInfo? {
-        let carrier: CTCarrier?
-        let radioTechnology: String?
+    init(networkInfo: CTTelephonyNetworkInfo) {
+        self.networkInfo = networkInfo
+        self.publisher = ValuePublisher(
+            initialValue: iOS12CarrierInfoProvider.readCarrierInfo(
+                from: networkInfo,
+                cellularProviderKey: networkInfo.serviceCurrentRadioAccessTechnology?.keys.first
+            )
+        )
 
-        if #available(iOS 12, *) {
-            guard let cellularProviderKey = networkInfo.serviceCurrentRadioAccessTechnology?.keys.first else {
-                return nil
+        // The `serviceSubscriberCellularProvidersDidUpdateNotifier` block object executes on the default priority
+        // global dispatch queue when the user’s cellular provider information changes.
+        // This occurs, for example, if a user swaps the device’s SIM card with one from another provider, while the app is running.
+        // ref.: https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo/3024512-servicesubscribercellularprovide
+        networkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { [weak self] cellularProviderKey in
+            guard let strongSelf = self else {
+                return
             }
-            radioTechnology = networkInfo.serviceCurrentRadioAccessTechnology?[cellularProviderKey]
-            carrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey]
-        } else {
-            radioTechnology = networkInfo.currentRadioAccessTechnology
-            carrier = networkInfo.subscriberCellularProvider
+
+            let carrierInfo = iOS12CarrierInfoProvider.readCarrierInfo(
+                from: strongSelf.networkInfo,
+                cellularProviderKey: cellularProviderKey
+            )
+
+            // On iOS12+ `CarrierInfo` subscribers are notified on actual change to cellular provider.
+            strongSelf.publisher.publishAsync(carrierInfo)
         }
+    }
 
-        guard let radioAccessTechnology = radioTechnology,
-            let currentCTCarrier = carrier else {
-                return nil
+    private static func readCarrierInfo(from networkInfo: CTTelephonyNetworkInfo, cellularProviderKey: String?) -> CarrierInfo? {
+        guard let cellularProviderKey = cellularProviderKey,
+           let radioTechnology = networkInfo.serviceCurrentRadioAccessTechnology?[cellularProviderKey],
+           let carrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey] else {
+            return nil // the service is not registered on any network
         }
+        return CarrierInfo(
+            carrierName: carrier.carrierName,
+            carrierISOCountryCode: carrier.isoCountryCode,
+            carrierAllowsVOIP: carrier.allowsVOIP,
+            radioAccessTechnology: .init(ctRadioAccessTechnologyConstant: radioTechnology)
+        )
+    }
```

Comment:
/nit I think you can simplify that by using an instance method to read the carrier, something like:

```swift
init(networkInfo: CTTelephonyNetworkInfo) {
    ...
    networkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { [weak self] cellularProviderKey in
        let carrierInfo = self?.readCarrierInfo(provider: cellularProviderKey)
        self?.publisher.publishAsync(carrierInfo)
    }
}

private func readCarrierInfo(provider key: String) -> CarrierInfo? {
    guard 
       let radioTechnology = networkInfo.serviceCurrentRadioAccessTechnology?[key],
       let carrier = networkInfo.serviceSubscriberCellularProviders?[key] 
    else {
       return nil // the service is not registered on any network
    }
    return CarrierInfo(
          carrierName: carrier.carrierName,
          carrierISOCountryCode: carrier.isoCountryCode,
          carrierAllowsVOIP: carrier.allowsVOIP,
          radioAccessTechnology: .init(ctRadioAccessTechnologyConstant: radioTechnology)
    )
}